### PR TITLE
Update ingress conformance test for finalizers

### PR DESCRIPTION
**Pull request by spencerhance**
**What type of PR is this?**
/kind failing-test

**What this PR does / why we need it**:
Updates Ingress Conformance tests to check for deletionTimestamp instead of 404 due to finalizers.

This will need to be cherry-picked to release-1.19

```release-note
Resolves an issue running Ingress conformance tests on clusters which use finalizers on Ingress objects to manage releasing load balancer resources
```

### CODE REVIEW

File: test/e2e/network/ingress.go [hunk 1]
```diff
@@ -1114,17 +1114,23 @@ var _ = SIGDescribe("Ingress API", func() {
 		ginkgo.By("deleting")
 		err = ingClient.Delete(context.TODO(), createdIngress.Name, metav1.DeleteOptions{})
 		framework.ExpectNoError(err)
-		_, err = ingClient.Get(context.TODO(), createdIngress.Name, metav1.GetOptions{})
-		framework.ExpectEqual(apierrors.IsNotFound(err), true, fmt.Sprintf("expected 404, got %#v", err))
+		ing, err := ingClient.Get(context.TODO(), createdIngress.Name, metav1.GetOptions{})
+		framework.ExpectNoError(err)
```
Comment (liggitt) [inline test/e2e/network/ingress.go:1118]: either of the following should be considered a success here:
1. apierrors.IsNotFound(err)
2. err == nil && ing.DeletionTimestamp != nil && len(ing.Finalizers) > 0

File: test/e2e/network/ingress.go [hunk 2]
```diff
@@ -1114,17 +1114,23 @@ var _ = SIGDescribe("Ingress API", func() {
...
+		framework.ExpectNotEqual(ing.DeletionTimestamp, nil, fmt.Sprint("expected deletionTimestamp, got nil"))
 		ings, err = ingClient.List(context.TODO(), metav1.ListOptions{LabelSelector: "special-label=" + f.UniqueName})
 		framework.ExpectNoError(err)
-		framework.ExpectEqual(len(ings.Items), 2, "filtered list should have 2 items")
+		// Should have 3 items since createdIngress will not have been deleted yet due to the finalizer
```
Comment (liggitt) [inline test/e2e/network/ingress.go:1122]: not all providers add finalizers... we can't assume this list length will be 3

File: test/e2e/network/ingress.go [hunk 3]
```diff
@@ -1114,17 +1114,23 @@ var _ = SIGDescribe("Ingress API", func() {
...
 		err = ingClient.DeleteCollection(context.TODO(), metav1.DeleteOptions{}, metav1.ListOptions{LabelSelector: "special-label=" + f.UniqueName})
 		framework.ExpectNoError(err)
 		ings, err = ingClient.List(context.TODO(), metav1.ListOptions{LabelSelector: "special-label=" + f.UniqueName})
 		framework.ExpectNoError(err)
-		framework.ExpectEqual(len(ings.Items), 0, "filtered list should have 0 items")
+		for _, ing := range ings.Items {
+			framework.ExpectNotEqual(ing.DeletionTimestamp, nil, fmt.Sprint("expected deletionTimestamp, got nil"))
+		}
+		// Should have 3 items since the ingresses will not have been deleted yet due to the finalizer
+		framework.ExpectEqual(len(ings.Items), 3, "filtered list should have 3 items")
```
Comment (liggitt) [inline test/e2e/network/ingress.go:1134]: we can't assume there will be a finalizer. we can assert the list length is <= 3, and that any existing items should have a deletionTimestamp and a finalizer

File: test/e2e/network/ingress.go (hunk 1 above)
Comment (spencerhance) [reply test/e2e/network/ingress.go]: Good call, done

File: test/e2e/network/ingress.go (hunk 2 above)
Comment (spencerhance) [reply test/e2e/network/ingress.go]: done

File: test/e2e/network/ingress.go (hunk 3 above)
Comment (spencerhance) [reply test/e2e/network/ingress.go]: done

**Comment by liggitt** [comment]
/lgtm

needs review by networking folks and conformance folks

**Comment by liggitt** [comment]
/ok-to-test

This looks like it would resolve the issues with the current test

**Comment by liggitt** [comment]
/lgtm cancel

looks like some verify errors and one issue with the e2e remain

File: test/e2e/network/ingress.go [hunk 4]
```diff
@@ -1114,17 +1114,35 @@ var _ = SIGDescribe("Ingress API", func() {
 		ginkgo.By("deleting")
 		err = ingClient.Delete(context.TODO(), createdIngress.Name, metav1.DeleteOptions{})
 		framework.ExpectNoError(err)
-		_, err = ingClient.Get(context.TODO(), createdIngress.Name, metav1.GetOptions{})
-		framework.ExpectEqual(apierrors.IsNotFound(err), true, fmt.Sprintf("expected 404, got %#v", err))
+		ing, err := ingClient.Get(context.TODO(), createdIngress.Name, metav1.GetOptions{})
+		// If ingress controller does not support finalizers, we expect a 404.  Otherwise we validate finalizer behavior.
+		if err != nil {
```
Comment (bowei) [inline test/e2e/network/ingress.go:1119]: Let's put the if err == nil case first.

Can we make this easier to read

```
ExpectFinalizer := func(ing *v1.Ingress) {
  framework.ExpectNotEqual(ing.DeletionTimestamp, nil, fmt.Sprint("expected deletionTimestamp, got nil"))
  framework.ExpectEqual(len(ing.Finalizers) > 0, true, fmt.Sprintf("expected finalizers on ingress, got none"))
}
...
```

use  `expectFinalizer(ing)` instead of copy/paste

Comment (spencerhance) [reply test/e2e/network/ingress.go]: done

**Comment by spencerhance** [comment]
Currently trying to resolve the test failures.  It looks like the deletionTimestamp is getting set but there are no finalizers.

@liggitt From looking at the logs, it seems that the apiserver is setting the deletionTimestamp, but there is no ingress controller in the conformance test to set the finalizer?

File: test/e2e/network/ingress.go [hunk 5]
```diff
@@ -1112,19 +1112,40 @@ var _ = SIGDescribe("Ingress API", func() {
 
 		// Ingress resource delete operations
 		ginkgo.By("deleting")
+
+		expectFinalizer := func(ing *networkingv1.Ingress) {
```
Comment (liggitt) [inline test/e2e/network/ingress.go:1116]: making this a helper method will make line numbers hard to follow when a failure happens... it won't be clear which step failed. suggest passing in a message or something so we can see which check failed

File: test/e2e/network/ingress.go [hunk 6]
```diff
@@ -1112,19 +1112,40 @@ var _ = SIGDescribe("Ingress API", func() {
...
+			framework.ExpectEqual(apierrors.IsNotFound(err), true, fmt.Sprintf("expected 404, got %v", err))
+		}
 		ings, err = ingClient.List(context.TODO(), metav1.ListOptions{LabelSelector: "special-label=" + f.UniqueName})
 		framework.ExpectNoError(err)
-		framework.ExpectEqual(len(ings.Items), 2, "filtered list should have 2 items")
+		// Should have <= 3 items since some ingresses might not have been deleted yet due to finalizers
+		framework.ExpectEqual(len(ings.Items) <= 3, true, "filtered list should have <= 3 items")
+		// Validate finalizers
+		for _, ing := range ings.Items {
+			expectFinalizer(&ing)
```
Comment (liggitt) [inline test/e2e/network/ingress.go:1136]: you only deleted `createdIngress.Name`, you should only expect a finalizer on that one:

```suggestion
for _, ing := range ings.Items {
	if ing.Namespace == createdIngress.Namespace && ing.Name == createdIngress.Name {
		// the ingress we deleted should only still exist if it has a finalizer
		expectFinalizer(&ing)
	}
}
```

File: test/e2e/network/ingress.go (hunk 5 above)
Comment (spencerhance) [reply test/e2e/network/ingress.go:1116]: Good call, updated

File: test/e2e/network/ingress.go (hunk 6 above)
Comment (spencerhance) [reply test/e2e/network/ingress.go]: Ah good catch 🤦‍♂️

File: test/e2e/network/ingress.go [hunk 7]
```diff
@@ -1112,19 +1112,42 @@ var _ = SIGDescribe("Ingress API", func() {
...
+		}
 		ings, err = ingClient.List(context.TODO(), metav1.ListOptions{LabelSelector: "special-label=" + f.UniqueName})
 		framework.ExpectNoError(err)
-		framework.ExpectEqual(len(ings.Items), 2, "filtered list should have 2 items")
+		// Should have <= 3 items since some ingresses might not have been deleted yet due to finalizers
+		framework.ExpectEqual(len(ings.Items) <= 3, true, "filtered list should have <= 3 items")
+		// Validate finalizer on the deleted ingress
+		for _, ing := range ings.Items {
+			if ing.Namespace == createdIngress.Namespace && ing.Name == createdIngress.Name {
+				expectFinalizer(&ing, "deleting all ingresses")
```
Comment (liggitt) [inline test/e2e/network/ingress.go:1137]:
```suggestion
				expectFinalizer(&ing, "listing after deleting createdIngress")
```

Comment (spencerhance) [reply test/e2e/network/ingress.go]: done

**Comment by spencerhance** [comment]
@liggitt looks like it just needs a milestone now

**Comment by liggitt** [comment]
@kubernetes/sig-release - would like a ruling on this for 1.20 vs 1.20.1. This should be merged and picked to 1.19 as well.

Without this PR, this conformance test (added in 1.19) fails on clusters that use finalizers on ingress objects to manage releasing load balancer resources.

**Review by hasheddan — APPROVED** [review]
/lgtm
/milestone v1.20

**Comment by liggitt** [comment]
Thanks. Picked to 1.19 in https://github.com/kubernetes/kubernetes/pull/96953
